### PR TITLE
SkyHelper-Networth v2

### LINKED ISSUE #49: Automated tests using Jest
Just wanted to write this down so I don't forget.

We can create a simple wrapper to make generating tests super simple with the handlers we have in the rewrite (#46). Basically, each test would be wrapped/decorated with a `HandlerTestWrapper` function or class or something (need to look into this). Then the test would only need to specify a few details including the handler class reference, test case inputs, whether the case should apply or not, and to price/calculation added.

Inspiration: https://github.com/pyinstaller/pyinstaller-hooks-contrib/blob/master/tests/test_libraries.py#L2170-L2174

Probably will do in #46

### LINKED ISSUE #53: Account for skins on dungeon heads
Like Knight Skin on Diamond Necron Head

### LINKED ISSUE #58: Base attribute type not accounted for
Right now, magic find tier 1 attribute price = blazing fortune tier 1 attribute price.

### LINKED ISSUE #59: No longer use winning bid for Midas Weapons
Because you can add coins later at [Zarina](https://wiki.hypixel.net/Zarina). AH seems to not care and we are setting values way lower. For example LBIN right now is 230M for a 2.46M Dark Auction buy price Midas Sword.

Image

### PR DESCRIPTION
**Pull request by DuckySoLucky**
Closes #14, closes #49, closes #53, closes #58, and closes #59.

## Highlights

- Non-cosmetic networth
- Rewrote codebase to be class-based using handlers for different price applications (e.g. recombobulator or enchantments)
- Settings and update managers
- Several new handlers
- Options to disable sorting and stacking
- Fixed numerous bugs and unexpected behavior
- Implemented automated testing and linting for QA

## New Usage

```js
const { ProfileNetworthCalculator, ItemNetworthCalculator, UpdateManager, NetworthManager, getPrices } = require('../../index');
const { getLatestProfile } = require('../functions');

(async () => {
    try {
        // Profile Networth Calculator
        const uuid = '6fd1d52fcb0f4d97b8d84542f93f2658';
        const { profile, profileData, museumData: profileMuseum } = await getLatestProfile(uuid, { museum: true });
        const bank = profileData.banking?.balance || 0;
        const museumData = profileMuseum[uuid];

        const networthManager = new ProfileNetworthCalculator(profile, museumData, bank);
        const networth = await networthManager.getNetworth({ onlyNetworth: false, includeItemData: true });

        const nonCosmeticNetworth = await networthManager.getNonCosmeticNetworth({ onlyNetworth: false, includeItemData: true });
        console.log(nonCosmeticNetworth.networth);
        console.log(networth.networth);

        // Item Networth Calculator
        const itemNetworthCalculator = new ItemNetworthCalculator(mockupItem);

        const itemNetwroth = await itemNetworthCalculator.getNetworth();
        const nonCosmeticItemNetworth = await itemNetworthCalculator.getNonCosmeticNetworth();
        console.log(nonCosmeticItemNetworth.price);
        console.log(itemNetwroth.price);

        // Networth Managers
        // ? Checks for updates of the package on npm.
        UpdateManager.checkForUpdate();

        // ? Changes the interval for checking for updates.
        UpdateManager.setInterval(1000 * 60 * 12 * 24); // 12 hours

        // ? Disables the interval for checking for updates.
        UpdateManager.disable();

        // ? Enables the interval for checking for updates.
        UpdateManager.enable();

        // ? Whether to cache the prices for 5 minutes after fetching them or fetch them every time.
        NetworthManager.setCachePrices(true);
        // ? Change amount of time to cache the prices in milliseconds. Default: 1000 * 60 * 5 (5 minutes)
        NetworthManager.setCachePricesTime(1000 * 60 * 5); // 5 minutes

        // ? Change whether to include item data in the networth calculation
        NetworthManager.setIncludeItemData(true);

        // ? Change interval between fetching items from Hypixel API
        NetworthManager.setItemsInterval(1000 * 60 * 60 * 12); // 12 hours

        // ? Change the number of retries when fetching items from Hypixel API
        NetworthManager.setItemsRetries(3);

        // ? Whether to only return the total networth or the items as well.
        NetworthManager.setOnlyNetworth(true);

        // ? Whether to stack items with the same attributes and price.
        NetworthManager.setStackItems(true);

        // Raw Prices
        const prices = await getPrices();
        console.log(prices['HYPERION']);
    } catch (error) {
        console.error(error);
    }
})();
```

### CODE REVIEW

File: calculators/items/Recombobulator_3000.js [hunk 1]
```diff
@@ -1,17 +1,10 @@
...
-        this.calculation = calculation;
-        this.itemLore = itemLore;
-        this.itemData = itemData;
-        this.prices = prices;
-        this.itemId = itemId;
-        this.price = price;
-        // this.calculate();
+class Recombobulator3000Calculation extends ItemHandler {
+    constructor() {
+        super();
```
Comment (MattTheCuber) [inline calculators/items/Recombobulator_3000.js]: No need to extend if you just call super()? Also why do we have data in the handlers? Why not static class methods that take in the item and prices?

**Comment by MattTheCuber** [comment]
What do you think about moving the `calculators/handlers` folder to root level?

File: calculators/items/Recombobulator_3000.js (hunk 1 above)
Comment (DuckySoLucky) [reply calculators/items/Recombobulator_3000.js]: Altpapier wanted it so we don't have to pass in any parameters in Handlers, so it just extends main handler class which has all parameters. Mostly just to keep the code cleaner I guess, cuz it's a lot of copy pasting, but the way you've done is good as well, passing in only item & prices, ask Altpapier I guess

**Comment by DuckySoLucky** [comment]
> What do you think about moving the `calculators/handlers` folder to root level?

Sure, idm

File: calculators/items/ItemCalculationHandler.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+        this.skyblockItem = getHypixelItemInformationFromId(this.itemId) ?? {};
+        this.itemName = this.itemData.tag.display.Name.replace(/§[0-9a-fk-or]/gi, '');
+        this.extraAttributes = this.itemData.tag.ExtraAttributes ?? {};
+        this.itemLore = this.itemData.tag.display.Lore ?? [];
+        this.count = this.itemData.Count ?? 1;
+        this.baseItemId = this.itemId;
+        this.calculation = calculation;
+        this.prices = prices;
+        this.price = price;
+    }
```
Comment (MattTheCuber) [inline calculators/items/ItemCalculationHandler.js]: This block of code should not have to be run for every handler in the calculation. All of these should only be computed once. That is why I removed the constructor before and moved to passing in arguments.

**Comment by MattTheCuber** [comment]
> > What do you think about moving the `calculators/handlers` folder to root level?
> 
> Sure, idm

https://discord.com/channels/@me/1239185792086310942/1275590837031534644

**Comment by MattTheCuber** [comment]
sorry if i messed up any of your work @DuckySoLucky

**Comment by Altpapier** [comment]
I added some notes in some files need some opinions on that. Also in the old item calculator we had prestiges calculated but on DeathStreeks profile it did not apply them ONCE bc u can now sell crimson or other armor on the ah. Is that used anywhere bc I removed it for now. Have not looked at pets yet bc that is also not correct yet but everything else *seems* be be working. Please also test out if everything is correct especially uppercase stuff

**Comment by Altpapier** [comment]
I also added a temporary itemsData.json file bc it seems like some1 removed the function that was used to get items idk why need explanation on that as well

**Comment by Altpapier** [comment]
and forgot to add: still need to implement nonCosmetic but need comment on the note in the helper file for items I left

File: calculators/ItemNetworthCalculator.js [hunk 3]
```diff
@@ -47,7 +45,7 @@ class ItemNetworthCalculator extends ItemNetworthHelper {
      */
     async #calculate(prices, { nonCosmetic, cachePrices, pricesRetries, includeItemData }) {
         // Set default values
-        this.nonCosmetic = nonCosmetic;
+        this.nonCosmetic = nonCosmetic; // Can be a bit confusing I thought the item was non cosmetic not the calculation. Suggestions?
```
Comment (MattTheCuber) [inline calculators/ItemNetworthCalculator.js]: change to `includeCosmetic` and default to true?

File: calculators/handlers.js
Comment (MattTheCuber) [inline calculators/handlers.js:1]: Good thinking XD

File: calculators/handlers/EssenceStars.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        const prestige = datas[0].type === 'prestige';
+        const calculationData = datas.reduce(
+            (acc, val) => {
+                acc.price += val?.price || 0;
+                return acc;
+            },
+            { id: prestigeItem, type: prestige ? 'prestige' : 'stars', price: 0, count: prestige ? 1 : star }
+        );
+
+        if (prestige && prices[prestigeItem.toLowerCase()]) calculationData.price += prices[prestigeItem.toLowerCase()];
```
Comment (MattTheCuber) [inline calculators/handlers/EssenceStars.js]: `toLowerCase`

**Review by MattTheCuber — COMMENTED** [review]
looks great! awesome work. this will be so much better to maintain. just a note, there has been a few additions to master since we started this PR that will need to get added.

**Comment by MattTheCuber** [comment]
i'll have to look at prestige stuff ltr.

File: calculators/handlers/EssenceStars.js (hunk 4 above)
Comment (Altpapier) [reply calculators/handlers/EssenceStars.js]: good point but if we dont need prestiges anymore it doesnt matter ;)

**Comment by MattTheCuber** [comment]
> Also in the old item calculator we had prestiges calculated but on DeathStreeks profile it did not apply them ONCE bc u can now sell crimson or other armor on the ah

i think we still need the prestige stuff because you can only sell the basic tier on the ah, not upgraded.

File: calculators/helpers/ItemNetworthHelper.js [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+            // if armor piece does not have base value
+            const prestige = prestiges[this.itemId.toUpperCase()];
+            if (prestige) {
+                for (const prestigeItem of prestige) {
+                    if (isNaN(this.price)) this.price = 0;
+                    if (this.skyblockItem?.upgrade_costs) this.price += starCosts(prices, this.calculation, this.skyblockItem.upgrade_costs, prestigeItem);
+                    if (this.skyblockItem?.prestige?.costs) this.price += starCosts(prices, this.calculation, this.skyblockItem.prestige.costs, prestigeItem);
+                }
+            }
+        }
```
Comment (MattTheCuber) [inline calculators/helpers/ItemNetworthHelper.js]: Should these be converted to handlers and the getBasePrice method runs a specific set of handlers that the calculator does not? Also noticed that god roll would be a base price, not added.

**Comment by MattTheCuber** [comment]
Just spent 30 minutes debugging to find that `gemstone_slots` is outdated in the old `items.json` for Dark Claymore...

**Comment by DuckySoLucky** [comment]
Tested on `MattTheCuber`, selected profile. everything matched with the old calculation. Haven't had time to test on other profiles / players. Gonna do that later. Some things were broken in the old networth calculation so they wouldn't match with the new one, so if you're gonna compare values, look for the comments with `? NOTE: USE THIS IF YOU'RE COMPARING OLD AND NEW`, and what it says, some require u to just uncomment it, while other tell u to remove specific part of the code fully. 

I compared items like this, you would just replace `types` category with the incorrect one which u can get from going thru `new.json` and `old.json` (I recommend u to turn on `onlyNetworth` param so it's easier to go thru it and compare values)
> Note: it's not perfect some items conflict (because it uses item.id for comparison, so if player has multiple of the same items with the same id in the same inventory they conflict) but good enough for temporary testing
> (didnt' test) Fix would be to most likely use index, but i'm not entirely sure if it's sorted same 
```js
const ProfileNetworthCalculator = require('../../calculators/ProfileNetworthCalculator');
const { getLatestProfile } = require('../functions');
const { getNetworth } = require('skyhelper-networth');
const fs = require('fs');

(async () => {
    try {
        const uuid = '40d803128602485382a112c60cfe84ae';
        const { profile, profileData, museumData: profileMuseum } = await getLatestProfile(uuid, { museum: true });

        const bank = profileData.banking?.balance || 0;
        const museumData = profileMuseum[uuid];
        const networthManager = new ProfileNetworthCalculator(profile, museumData, bank);
        const networth = await networthManager.getNetworth(null, { onlyNetworth: false });
        const old = await getNetworth(profile, bank, {
            museumData: museumData,
            onlyNetworth: false,
            v2Endpoint: true,
            cache: true,
        });

        console.log(old);
        console.log(networth);
        fs.writeFileSync('old.json', JSON.stringify(old, null, 4));
        fs.writeFileSync('new.json', JSON.stringify(networth, null, 4));

        fs.writeFileSync('oldNetworth.json', JSON.stringify(old.types.storage, null, 4));
        fs.writeFileSync('newNetworth.json', JSON.stringify(networth.types.storage, null, 4));

        const oldE = old.types.storage.items;
        const newE = networth.types.storage.items;
        const differences = oldE.filter((item) => {
            const oldItem = item;
            const newItem = newE.find((i) => i.id.toLowerCase() === item.id);

            if (oldItem.price !== newItem.price) {
                console.log(oldItem.id, newItem.id);
            }

            return oldItem.price !== newItem.price;
        });

        fs.writeFileSync('differences.json', JSON.stringify(differences, null, 4));
        console.log('done', differences.length);

        console.log(oldE.find((i) => i.id === 'epic_turtle').price); //.filter((i) => i.type === 'enchant'));
        console.log(newE.find((i) => i.id === 'EPIC_TURTLE').price); //.filter((i) => i.type === 'ENCHANT'));
    } catch (error) {
        console.error(error);
    }
})();
```

**Comment by DuckySoLucky** [comment]
Tested on DeathStreeks' main profile, everything matches. The only difference is dead axe with the `GOLDEN_BOUNTY` (scaveger 6+).
 The only broken part were pets outside of the pets inventory, I made a [fix](https://github.com/Altpapier/SkyHelper-Networth/pull/46/files#diff-b4ac643dc3cf19600b26ffa343f3131477831d99317c8358a4c56d1c78552f43R158-R165), but it's not really great, probably should be done differently, any suggestions?



Tested on 56ms' 1T profile, everything matches (with those toggles)

**Comment by Altpapier** [comment]
what is the current progress on this? what is there still left to do (except tests)?

**Comment by MattTheCuber** [comment]
> what is the current progress on this? what is there still left to do (except tests)?

Auto tests can be done later. I don't want it to hold up progress so I made an issue for it.

Other than that we need to:

- Finish writing the manager singletons (auto `/items` updater and auto `prices.json` updater)
- ~~Fully manually test the profile networth calculator and compare to old results on many profiles (including high-value, low-value, and edge case profiles)~~
- Update `README.md` and examples
- Random cleanup like removing `OLD_itemCalculator.js` and other unused functions

**Comment by MattTheCuber** [comment]
Profiles I've tested with perfect match:

- MattTheCuber on Strawberry
- MattTheCuber on Apple
- MattTheCuber on Grapes
- Altpapier on Watermelon
- DuckySoSkilled on Raspberry
- 56ms on Strawberry
- Refraction on Apple
- Deathstreeks on Blueberry
- juggfein on Grapes
- InqWhisperer on Mango
- vaporer on Strawberry
- LeaPhant on Grapes
- Akinsoft on Not Allowed To Quit Skyblock Ever Again
- Dueces on Kiwi
- TheOriginalAce on Mango
- Chissl on Lime
- HuTayo on Mango
- builder_247 on Tomato
- Ealman on Papaya
- Eisengolem on Kiwi
- Eisengolem on Banana

https://github.com/HypixelDatabase/HypixelTracking/blob/master/urls.js#L54-L66

**Comment by Altpapier** [comment]
> > what is the current progress on this? what is there still left to do (except tests)?
> 
> Auto tests can be done later. I don't want it to hold up progress so I made an issue for it.
> 
> Other than that we need to:
> 
>     * Finish writing the manager singletons (auto `/items` updater and auto `prices.json` updater)
> 
>     * ~Fully manually test the profile networth calculator and compare to old results on many profiles (including high-value, low-value, and edge case profiles)~
> 
>     * Update `README.md` and examples
> 
>     * Random cleanup like removing `OLD_itemCalculator.js` and other unused functions

once we are done with that I would also suggest us implementing these changes in another branch in the bot so that we can find some other bugs before publishing this version

File: calculators/BasicItemNetworthCalculator.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+     * Creates a new BasicItemNetworthCalculator
+     * @param {object} itemData The item the networth should be calculated for
+     * @param {string} itemData.id The item's id
+     * @param {number} itemData.amount The amount of the item
+     */
+    constructor({ id, amount }) {
+        this.id = id;
+        this.amount = amount;
+        this.skyblockItem = getHypixelItemInformationFromId(this.id) ?? {};
+        this.itemName = this.skyblockItem?.name || titleCase(this.id).split(' ').reverse().join(' '); //TODO: i think this was only needed for essence?
```
Comment (Altpapier) [inline calculators/BasicItemNetworthCalculator.js]: todo

File: calculators/ProfileNetworthCalculator.js [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+                if (!onlyNetworth && result && result?.price) categories[category].items.push(result);
+            }
+
+            // Sort items by price
+            if (!onlyNetworth && categories[category].items.length > 0) {
+                categories[category].items = categories[category].items.sort((a, b) => b.price - a.price);
+
+                // Stack items with the same name and price
+                if (stackItems) {
+                    // TODO: broken for 56ms fishing_bag (not broken before classes update)
```
Comment (Altpapier) [inline calculators/ProfileNetworthCalculator.js]: todo

File: calculators/handlers/GemstonePowerScroll.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        return item.extraAttributes.power_ability_scroll;
+    }
+
+    /**
+     * Calculates and adds the price of the modifier to the item
+     * @param {object} item The item data
+     * @param {object} prices A prices object generated from the getPrices function
+     */
+    calculate(item, prices) {
+        // TODO: Remove This
```
Comment (Altpapier) [inline calculators/handlers/GemstonePowerScroll.js]: todo

File: calculators/handlers/Gemstones.js [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+                        gemstoneSlots.splice(slotIndex, 1);
+                    }
+                }
+            }
+        }
+
+        // GEMSTONES
+        for (const gemstone of gems) {
+            // TODO: Remove This
```
Comment (Altpapier) [inline calculators/handlers/Gemstones.js]: todo

File: calculators/handlers/ItemEnchantments.js [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+                        count: efficiencyLevel,
+                    };
+                    item.price += calculationData.price;
+                    item.calculation.push(calculationData);
+                }
+            }
+
+            // GOLDEN BOUNTY (scavenger 6+: https://hypixel.net/threads/scavenger-7-now-obtainable.5741331/, NEED MORE INFO! before: value === 6)
+
+            // TODO: Remove This
```
Comment (Altpapier) [inline calculators/handlers/ItemEnchantments.js]: todo

File: calculators/helpers/PetNetworthHelper.js [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+        this.base = this.price;
+    }
+
+    /**
+     * Gets the pet level
+     * @returns {object} The pet level
+     */
+    getPetLevel() {
+        // TODO: Move from constants to helper
```
Comment (Altpapier) [inline calculators/helpers/PetNetworthHelper.js]: todo

**Review by Altpapier — COMMENTED** [review]
I have added some more comments to all the left todo comments I found in the code. maybe some1 can have a look and see if any of those still needs to be done or what there is left to do with them

File: calculators/BasicItemNetworthCalculator.js (hunk 6 above)
Comment (DuckySoLucky) [reply calculators/BasicItemNetworthCalculator.js]: It's also used for `Beastmaster Crest`, `Griffin Upgrade Stone`, `Wisp Upgrade Stone` & `Exp Boost` (Obtaining their rarity)

File: calculators/ProfileNetworthCalculator.js (hunk 7 above)
Comment (DuckySoLucky) [reply calculators/ProfileNetworthCalculator.js]: Cannot reproduce

**Comment by DuckySoLucky** [comment]
https://github.com/Altpapier/SkyHelper-Networth/pull/50

**Comment by MattTheCuber** [comment]
> > > what is the current progress on this? what is there still left to do (except tests)?
> > 
> > 
> > Auto tests can be done later. I don't want it to hold up progress so I made an issue for it.
> > Other than that we need to:
> > ```
> > * Finish writing the manager singletons (auto `/items` updater and auto `prices.json` updater)
> > 
> > * ~Fully manually test the profile networth calculator and compare to old results on many profiles (including high-value, low-value, and edge case profiles)~
> > 
> > * Update `README.md` and examples
> > 
> > * Random cleanup like removing `OLD_itemCalculator.js` and other unused functions
> > ```
> 
> once we are done with that I would also suggest us implementing these changes in another branch in the bot so that we can find some other bugs before publishing this version

What are you referring to? Implementing what?

File: calculators/handlers/Prestige.js [hunk 12]
```diff
@@ -24,7 +23,7 @@ class PrestigeHandler {
         if (prices[item.itemId]) return;
         const prestige = prestiges[item.itemId];
         for (const prestigeItem of prestige) {
-            const foundItem = getHypixelItemInformationFromId(prestigeItem);
+            const foundItem = item?.skyblockItem;
```
Comment (MattTheCuber) [inline calculators/handlers/Prestige.js]: This should be reverted. This is getting the item information for a different item than is being calculating. `prestigeItem !== item.itemId`

**Review by MattTheCuber — CHANGES_REQUESTED** [review]

File: calculators/handlers/Prestige.js (hunk 12 above)
Comment (Altpapier) [reply calculators/handlers/Prestige.js]: oops didn't see

**Comment by Altpapier** [comment]
> What are you referring to? Implementing what?

Adding another branch to the SkyHelper repo where we implement the new version of skyhelper-networth and use the "new" package in our bot first before publishing the release of this

**Comment by MattTheCuber** [comment]
TODO: Update license to GPLv3

**Comment by DuckySoLucky** [comment]
All methods & fields got full documentation, from explanation to the type

```js
const { ProfileNetworthCalculator, ItemNetworthCalculator, UpdateManager, NetworthManager, getPrices } = require('../../index');
const { getLatestProfile } = require('../functions');

(async () => {
    try {
        // Profile Networth Calculator
        const uuid = '6fd1d52fcb0f4d97b8d84542f93f2658';
        const { profile, profileData, museumData: profileMuseum } = await getLatestProfile(uuid, { museum: true });
        const bank = profileData.banking?.balance || 0;
        const museumData = profileMuseum[uuid];

        const networthManager = new ProfileNetworthCalculator(profile, museumData, bank);
        const networth = await networthManager.getNetworth({ onlyNetworth: false, includeItemData: true });

        const nonCosmeticNetworth = await networthManager.getNonCosmeticNetworth({ onlyNetworth: false, includeItemData: true });
        console.log(nonCosmeticNetworth.networth);
        console.log(networth.networth);

        // Item Networth Calculator
        const itemNetworthCalculator = new ItemNetworthCalculator(mockupItem);

        const itemNetwroth = await itemNetworthCalculator.getNetworth();
        const nonCosmeticItemNetworth = await itemNetworthCalculator.getNonCosmeticNetworth();
        console.log(nonCosmeticItemNetworth.price);
        console.log(itemNetwroth.price);

        // Networth Managers
        // ? Checks for updates of the package on npm.
        UpdateManager.checkForUpdate();

        // ? Changes the interval for checking for updates.
        UpdateManager.setInterval(1000 * 60 * 12 * 24); // 12 hours

        // ? Disables the interval for checking for updates.
        UpdateManager.disable();

        // ? Enables the interval for checking for updates.
        UpdateManager.enable();

        // ? Whether to cache the prices for 5 minutes after fetching them or fetch them every time.
        NetworthManager.setCachePrices(true);
        // ? Change amount of time to cache the prices in milliseconds. Default: 1000 * 60 * 5 (5 minutes)
        NetworthManager.setCachePricesTime(1000 * 60 * 5); // 5 minutes

        // ? Change whether to include item data in the networth calculation
        NetworthManager.setIncludeItemData(true);

        // ? Change interval between fetching items from Hypixel API
        NetworthManager.setItemsInterval(1000 * 60 * 60 * 12); // 12 hours

        // ? Change the number of retries when fetching items from Hypixel API
        NetworthManager.setItemsRetries(3);

        // ? Whether to only return the total networth or the items as well.
        NetworthManager.setOnlyNetworth(true);

        // ? Whether to stack items with the same attributes and price.
        NetworthManager.setStackItems(true);

        // Raw Prices
        const prices = await getPrices();
        console.log(prices['HYPERION']);
    } catch (error) {
        console.error(error);
    }
})();
```

File: calculators/ItemNetworthCalculator.js [hunk 13]
```diff
@@ -3,34 +3,36 @@ const ItemNetworthHelper = require('./helpers/ItemNetworthHelper');
...
- * Class for calculating the networth of an item
+ * @typedef {import('../types/ItemNetworthCalculator').ItemNetworthCalculator} ItemNetworthCalculator
+ * @typedef {import('../types/ItemNetworthCalculator').Item} NetworthResult
+ * @typedef {import('../types/global').NetworthOptions} NetworthOptions
+ */
+
+/**
+ * ItemNetworthCalculator class.
+ * Calculates the networth of an item.
+ * @implements {ItemNetworthCalculator}
```
Comment (MattTheCuber) [inline calculators/ItemNetworthCalculator.js]:
```suggestion
 * @implements {ItemNetworthHelper}
```
?

File: managers/NetworthManager.js [hunk 14]
```diff
@@ -4,36 +4,47 @@ const { sleep } = require('../helper/functions');
...
+    #itemsInterval;
+    #onlyNetworth;
+    #stackItems;
+    #includeItemData;
+    #itemsIntervalInstance;
     /**
      * Create a new NetworthManager instance. This class is a singleton and should be accessed through the networthManager instance
      * @param {Object} options - Options for the NetworthManager
-     * @param {boolean | number} [options.cachePrices=true] - Whether to cache the prices for 5 minutes after fetching them or fetch them every time. Can also be a number to cache the prices for a specific amount of time in milliseconds
+     * @param {boolean | number} [options.cachePrices=true] - Whether to cache the prices for 5 minutes after fetching them or fetch them every time.
```
Comment (MattTheCuber) [inline managers/NetworthManager.js]:
```suggestion
     * @param {boolean | number} [options.cachePrices=true] - Whether to cache the prices for a time after fetching them or fetch them every time.
```

File: managers/NetworthManager.js [hunk 15]
```diff
@@ -4,36 +4,47 @@ const { sleep } = require('../helper/functions');
...
 class NetworthManager {
+    #cachePrices;
+    #cachePricesTime;
+    #pricesRetries;
+    #itemsRetries;
+    #itemsInterval;
+    #onlyNetworth;
+    #stackItems;
+    #includeItemData;
+    #itemsIntervalInstance;
```
Comment (MattTheCuber) [inline managers/NetworthManager.js]: Why?

File: constants/misc.js [hunk 16]
```diff
@@ -98,18 +79,30 @@ const specialEnchantmentMatches = {
...
+const NON_COSMETIC_ITEMS = new Set([
+    'DCTR_SPACE_HELM',
+    'DCTR_SPACE_HELM_EDITIONED',
+    'DEAD_BUSH_OF_LOVE',
+    'ANCIENT_ELEVATOR',
+    'CREATIVE_MIND_UNEDITIONED',
+    'DUECES_BUILDER_CLAY',
+    'CREATIVE_MIND',
+    'GAME_BREAKER',
+    'POTATO_BASKET',
```
Comment (MattTheCuber) [inline constants/misc.js]:
```suggestion
    'ANCIENT_ELEVATOR',
    'BEDROCK',
    'CREATIVE_MIND',
    'CREATIVE_MIND_UNEDITIONED',
    'DCTR_SPACE_HELM',
    'DCTR_SPACE_HELM_EDITIONED',
    'DEAD_BUSH_OF_LOVE',
    'DUECES_BUILDER_CLAY',
    'GAME_BREAKER',
    'POTATO_BASKET',
```

File: constants/pets.js [hunk 17]
```diff
@@ -16,10 +24,10 @@ const LEVELS = [
     1886700, 1886700, 1886700, 1886700, 1886700, 1886700, 1886700, 1886700,
 ];
 
-const BLOCKED_CANDY_REDUCE_PETS = ['ENDER_DRAGON', 'GOLDEN_DRAGON', 'SCATHA'];
-const RECOMB_PET_ITEMS = ['PET_ITEM_TIER_BOOST'];
-const SOULBOUND_PETS = ['GRANDMA_WOLF', 'KUUDRA', 'BINGO'];
-const TIERS = ['COMMON', 'UNCOMMON', 'RARE', 'EPIC', 'LEGENDARY', 'MYTHIC', 'DIVINE', 'SPECIAL', 'VERY_SPECIAL', 'ULTIMATE'];
+const CUSTOM_PET_NAMES = {
+    TYRANNOSAURUS: 'T-Rex',
+    FRACTURED_MONTEZUMA_SOUL: 'Montezuma',
```
Comment (MattTheCuber) [inline constants/pets.js:27]: Did we fix this in main bot?

File: calculators/helpers/ItemNetworthHelper.js [hunk 18]
```diff
@@ -126,10 +132,11 @@ class ItemNetworthHelper {
 
     /**
      * Checks if the item is a rune
+     * @param {boolean} [allRunes=false] Whether to include all runes or just unique runes
      * @returns {boolean} Whether the item is a rune
      */
-    isRune() {
-        const isRuneId = this.itemId === 'RUNE' || this.itemId === 'UNIQUE_RUNE';
+    isRune(allRunes = false) {
```
Comment (MattTheCuber) [inline calculators/helpers/ItemNetworthHelper.js]: Generally its better to make 2 separate methods to avoid confusion when calling `this.isRune(true)` where instead it could be `this.isRune()` and `this.isUniqueRune()` or something.

File: calculators/helpers/PetNetworthHelper.js [hunk 19]
```diff
@@ -59,6 +61,14 @@ class PetNetworthHelper {
         return SOULBOUND_PETS.includes(this.petData.type);
     }
 
+    /**
+     * Checks if the pet is cosmetic
+     * @returns {boolean} Whether the pet is cosmetic
+     */
+    isCosmetic() {
+        return !!this.skin;
+    }
```
Comment (MattTheCuber) [inline calculators/helpers/PetNetworthHelper.js:94]: Well, the whole pet shouldn't be excluded... Just the skin part of the pet should be excluded.

File: calculators/helpers/PetNetworthHelper.js [hunk 20]
```diff
@@ -119,14 +129,19 @@ class PetNetworthHelper {
         this.base = this.price;
     }
 
+    getPetId(prices, nonCosmetic) {
```
Comment (MattTheCuber) [inline calculators/helpers/PetNetworthHelper.js:159]: Missing jsdoc

File: calculators/handlers/MidasWeapon.js [hunk 21]
```diff
@@ -19,8 +19,20 @@ class MidasWeaponHandler {
...
+        } else if (item.itemId === 'starred_midas_sword') {
+            maxBid = 250_000_000;
+            type = 'starred_midas_sword_250m';
+        } else if (item.itemId === 'midas_staff') {
+            maxBid = 100_000_000;
+            type = 'midas_staff_100m';
+        } else if (item.itemId === 'starred_midas_staff') {
+            maxBid = 500_000_000;
+            type = 'starred_midas_staff_500m';
+        }
```
Comment (MattTheCuber) [inline calculators/handlers/MidasWeapon.js]: Caps?

File: calculators/ProfileNetworthCalculator.js [hunk 22]
```diff
@@ -13,26 +13,30 @@ const categoryCalculatorMap = {
...
- * Class for calculating the networth of a profile
+ * @typedef {import('../types/ProfileNetworthCalculator').ProfileNetworthCalculator} ProfileNetworthCalculator
+ * @typedef {import('../types/ProfileNetworthCalculator').NetworthResult} NetworthResult
+ * @typedef {import('../types/global').NetworthOptions} NetworthOptions
+ */
+
+/**
+ * ProfileNetworthCalculator class.
+ * Calculates the networth of a player's profile.
+ * @implements {ProfileNetworthCalculator}
```
Comment (MattTheCuber) [inline calculators/ProfileNetworthCalculator.js:27]:
```suggestion
```
?

**Review by MattTheCuber — CHANGES_REQUESTED** [review]
Found a couple minor things and potentially a bug or 2

File: managers/NetworthManager.js (hunk 15 above)
Comment (MattTheCuber) [reply managers/NetworthManager.js]: > to hide those settings from user

> there r getters and setters

> so user knows what to put

> and what he's gonna get

> and all that

**Comment by MattTheCuber** [comment]
## Status update
Currently being tested with production SkyHelper bot.

File: calculators/GenericItemNetworthCalculator.js [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+/**
+ * @typedef {import('../types/ItemNetworthCalculator').ItemNetworthCalculator} ItemNetworthCalculator
+ * @typedef {import('../types/ItemNetworthCalculator').Item} NetworthResult
+ * @typedef {import('../types/global').NetworthOptions} NetworthOptions
+ */
+
+/**
+ * ItemNetworthCalculator class.
+ * Calculates the networth of an item.
+ * @implements {ItemNetworthCalculator}
```
Comment (MattTheCuber) [inline calculators/GenericItemNetworthCalculator.js]: This intended?

File: calculators/GenericItemNetworthCalculator.js [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+         */
+
+        if (this.item.tag?.ExtraAttributes?.petInfo || this.item.exp !== undefined) {
+            try {
+                this.item = this.item.tag?.ExtraAttributes?.petInfo ? JSON.parse(this.item.tag.ExtraAttributes.petInfo) : this.item;
+                calculatorClass = PetNetworthCalculator;
+            } catch {}
+        } else if (!this.item.tag?.ExtraAttributes && this.item.exp === undefined && typeof this.item.id !== 'string') {
+            throw new ValidationError('Invalid item data.');
+        }
```
Comment (MattTheCuber) [inline calculators/GenericItemNetworthCalculator.js]: Where `BasicItemNetworthCalculator`?

**Comment by MattTheCuber** [comment]
Details on the fix in the last commit:

### Before

```
Calculations:
→ Fiery Fervor Leggings Prestige (31.23M)
→ 10x Fiery Fervor Leggings Stars (109.45M)
→ Burning Fervor Leggings Prestige (15.96M) // Included base item price incorrectly
→ 10x Burning Fervor Leggings Stars (22.05M)
→ Hot Fervor Leggings Prestige (8.2M)  // Included base item price incorrectly
→ 10x Hot Fervor Leggings Stars (6.3M)
→ Fervor Leggings Prestige (1.3M)  // Included base item price correctly
→ 10x Fervor Leggings Stars (2.67M)
```

```
Calculations:
→ Fiery Fervor Leggings Prestige (31.23M)
→ 10x Fiery Fervor Leggings Stars (109.45M)
→ Burning Fervor Leggings Prestige (5.96M) // No longer includes base item price
→ 10x Burning Fervor Leggings Stars (22.05M)
→ Burning Fervor Leggings Prestige Base Item (10M) // Instead it's on its own
// Short circuits since we don't need to compute the price any further down the line
```

**Review by Altpapier — APPROVED** [review]
